best-practices-cel/disallow-default-namespace/disallow-default-namespace.yaml

```diff
@@ -0,0 +1,50 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: disallow-default-namespace
+  annotations:
+    pod-policies.kyverno.io/autogen-controllers: none
+    policies.kyverno.io/title: Disallow Default Namespace in CEL expressions
+    policies.kyverno.io/minversion: 1.11.0
+    policies.kyverno.io/category: Multi-Tenancy in CEL 
+    kyverno.io/kubernetes-version: "1.26-1.27"
+    policies.kyverno.io/severity: medium
+    policies.kyverno.io/subject: Pod
+    policies.kyverno.io/description: >-
+      Kubernetes Namespaces are an optional feature that provide a way to segment and
+      isolate cluster resources across multiple applications and users. As a best
+      practice, workloads should be isolated with Namespaces. Namespaces should be required
+      and the default (empty) Namespace should not be used. This policy validates that Pods
+      specify a Namespace name other than `default`. Rule auto-generation is disabled here
+      due to Pod controllers need to specify the `namespace` field under the top-level `metadata`
+      object and not at the Pod template level.
+spec:
+  validationFailureAction: Audit
+  background: true
+  rules:
+  - name: validate-namespace
+    match:
+      any:
+      - resources:
+          kinds:
+          - Pod
+    validate:
+      cel:
+        expressions:
+          - expression: "object.metadata.namespace != 'default'"
```

Comment:
Applying this policy produces a warning:
```
Warning: spec.rules[0].validate.cel.expressions[0].expression:/v1, Kind=Pod: ERROR: <input>:1:16: undefined field 'namespace'; | object.metadata.namespace != 'default'; | ...............^;
clusterpolicy.kyverno.io/disallow-default-namespace created
```

The corresponding VAP produces the same warning. It is better to access the namespace name using the `namespaceObject` instead.

It would be something like this:
```
expression: "namespaceObject.metadata.name != 'default'"
```

For more info, check [this](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#validation-expression) doc.

Reply:
I have made the changes, but when we run the kyverno test we get a `no such key: metadata` error:
```
$ kyverno test . --detailed-results

Loading test  ( kyverno-test.yaml ) ...
  Loading values/variables ...
  Loading policies ...
  Loading resources ...
  Applying 1 policy to 4 resources ...
  Checking results ...

│────│────────────────────────────│──────────────────────────────────│─────────────────────────────│────────│─────────────────────│──────────────────────────────────────────────────────────────────────────────────────────────────│
│ ID │ POLICY                     │ RULE                             │ RESOURCE                    │ RESULT │ REASON              │ MESSAGE                                                                                          │
│────│────────────────────────────│──────────────────────────────────│─────────────────────────────│────────│─────────────────────│──────────────────────────────────────────────────────────────────────────────────────────────────│
│ 1  │ disallow-default-namespace │ validate-namespace               │ Pod/badpod01                │ Pass   │ Ok                  │ expression 'namespaceObject.metadata.name != 'default'' resulted in error: no such key: metadata │
```
The kyverno test has to be removed due to this error. 
Shall I log an issue to support `namespaceObject` variable in CLI tests?

Reply:
Good catch! 
Yeah go ahead and file an issue.

Reply:
I created the issue. I will work on it tomorrow.

Reply:
Thanks for the help 🙌

Reply:
I will add the kyverno tests once the PR to fix it merged